### DISCUSSION TITLE
Proposal to support `ipnet` crate

### DISCUSSION THREAD

**Discussion by bnaecker** [diesel-rs/diesel#3164] [Ideas]
Diesel currently supports the [`ipnetwork`](https://docs.rs/ipnetwork/latest/ipnetwork/) crate for working with IP addresses and subnets in CIDR block notation. I'd like to propose an additional feature, allowing consumers to use the [`ipnet`](https://docs.rs/ipnet/2.5.0/ipnet/struct.Ipv4Net.html) as well. The second crate presents a different API, such as facilities for managing subnetworks and traits for bitwise operations on addresses. `ipnet` is also significantly more popular than `ipnetwork`, though that may be a secondary concern.

I'd be happy to provide additional background on either crate, and note that we'd be more than willing to implement the proposed feature. Thanks!

**Comment by smklein** [comment]
One possible prototype of this would be here: https://github.com/diesel-rs/diesel/pull/3165

**Comment by weiznich** [comment]
I'm generally fine with adding support for a popular third party crate. That written, I would like to see some discussion around the fact if this support should be part of diesel or if it should be the other way around and this should be added to `ipnet` because that the other location such support can be provided.

**Reply by smklein** [reply]
Seems reasonable - do you have a sense of what qualifies a type as "this should have the ToSql/FromSql implementation in Diesel" vs "this should have the ToSql/FromSql implementation in the third party crate"?

My intuition with the prototype was "in Diesel" partially because that aligned with the existing `ipnetwork` implementation.

My (limited!) opinion:
- If a crate is more popular + stable, it makes sense for the implementations to live in Diesel. This keeps popular crates relatively small, and the change is hopefully unobtrusive/low-maintenance in Diesel itself. Diesel's feature flag support also already has a precedent, so it's easy to select the "flavor" of Diesel desired.
- If a crate is less popular + less stable, it makes sense for the implementations to live in the remote crate. This keeps the maintenance burden on "whoever is doing the churning".

**Reply by weiznich** [reply]
These points summarize the decision making quite well, but I think there is another relevant point here:
* How many users of diesel/the other crate would benefit from the added support. To give some examples for this: If the other crate is basically everywhere, including a lot of contexts where users may not use diesel it's probably reasonable to include support for this crate in diesel itself. On the other hand if the other crate is used almost exclusively with diesel it's reasonable to move the "diesel integration" to the other crate.

EDIT: That's likely something we should also ask the maintainers of `ipnet`. Mind opening an issue there?

**Reply by smklein** [reply]
> If the other crate is basically everywhere, including a lot of contexts where users may not use diesel it's probably reasonable to include support for this crate in diesel itself.

I'm not sure an issue is required to answer that question - https://crates.io/crates/ipnet shows that ipnet has 20,000,000+ downloads, and it's one of the [top ten most popular networking crates across crates.io](https://crates.io/categories/network-programming). I can say with high confidence that is it not being used exclusively with Diesel (which has 3,500,000+ downloads on crates.io).

**Reply by weiznich** [reply]
That's by itself not a argument in my opinion. For example [rust_decimal](https://lib.rs/crates/rust_decimal) is also a crate that seems to be more popular than diesel by itself. Nevertheless they provide diesel support behind a feature flag.

**Reply by smklein** [reply]
Filed: https://github.com/krisprice/ipnet/issues/42
